group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2863,6 +2864,93 @@ private void removePendingSyncMember(
         }
     }
 
+    /**
+     * Handle a generic group HeartbeatRequest.
+     *
+     * @param context        The request context.
+     * @param request        The actual Heartbeat request.
+     *
+     * @return The Heartbeat response.
+     */
+    public HeartbeatResponseData genericGroupHeartbeat(
+        RequestContext context,
+        HeartbeatRequestData request
+    ) {
+        String groupId = request.groupId();
+        String memberId = request.memberId();
+        String groupInstanceId = request.groupInstanceId();
+        int generationId = request.generationId();
+        GenericGroup group;
+        try {
+            group = getOrMaybeCreateGenericGroup(groupId, false);
+        } catch (Throwable t) {
+            return new HeartbeatResponseData()
+                .setErrorCode(Errors.forException(t).code());
+        }
+
+        Optional<Errors> errorOpt = validateGenericGroupHeartbeat(group, memberId, groupInstanceId, generationId);
+        if (errorOpt.isPresent()) {
+            return new HeartbeatResponseData()
+                .setErrorCode(errorOpt.get().code());
+        } else {
+            switch (group.currentState()) {
+                case EMPTY:
+                    return new HeartbeatResponseData().setErrorCode(Errors.UNKNOWN_MEMBER_ID.code());
+                case COMPLETING_REBALANCE:
+                    // Consumers may start sending heartbeat after join-group response, in which case
+                    // we should treat them as normal hb request and reset the timer
+                    GenericGroupMember member = group.member(memberId);
```

Comment:
nit: The fact that member is declared only here is a bit weird. To avoid it, we could rewrite the code like this:

```
rescheduleGenericGroupMemberHeartbeat(
    group,
    group.member(memberId)
);
```